torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.cpp

```diff
@@ -3,14 +3,166 @@
 #include <torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.h>
 #include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
 
+#include "glfdc/eval.h"
+#include "glfdc/expr_builder.h"
+
 #include <iostream>
 
+namespace {
+
+namespace fuser_cuda = torch::jit::fuser::cuda;
+using namespace torch::jit::fuser::cuda::kir;
+
+class ExprBuilderVisitor : IrVisitor {
+ public:
+  explicit ExprBuilderVisitor(glfdc::ExpressionBuilder& bld);
+
+  c10::optional<glfdc::Expr> build(const Val* value) {
+    auto ref = build_operand(value);
+    if (!glfdc::is_sexpr(ref))
+      return c10::nullopt;
+    return glfdc::Expr{&builder_->dag(), c10::get<glfdc::SExprRef>(ref)};
+  }
+
+  glfdc::Operand build_operand(const Val* value) {
+    TORCH_CHECK(value->isScalar());
+    TORCH_CHECK(value->dtype() == fuser_cuda::DataType::Int);
+
+    if (value->isA<Int>()) {
+      if (value->isConst()) {
+        auto opt_val = value->as<Int>()->value();
+        if (opt_val.has_value())
+          return glfdc::Value(opt_val.value());
+        else
+          return builder_->get_binding(reinterpret_cast<uintptr_t>(value));
+      }
+    }
+
+    current_operand_ = c10::nullopt;
+    value->accept(this);
+    TORCH_CHECK(current_operand_.has_value());
+    auto ret = current_operand_.value();
+    current_operand_ = c10::nullopt;
+    return ret;
+  }
+
+ private:
+  void unhandled(const void*) final;
+  void visit(const Int* value) final;
+  void visit(const NamedScalar* named_scalar) final;
+  void visit(const UnaryOp* unary_op) final;
+  void visit(const BinaryOp* binary_op) final;
+
+  glfdc::ExpressionBuilder* builder_;
```

Comment:
nit: `builder_ = nullptr`, or make it a reference.

Reply:
This code is removed